src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,196 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    public Node<T> getFirst() {
+        return first;
+    }
+
+    public Node<T> getLast() {
+        return last;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private void addFirst(T t) {
+        Node<T> newNode;
+        if (first == null) {
+            newNode = new Node<>(null, t, null);
+            last = newNode;
+        } else {
+            newNode = new Node<>(null, t, first);
+            first.prev = newNode;
+        }
+        first = newNode;
+        size++;
+    }
+
+    private void addLast(T t) {
+        Node<T> newNode;
+        if (last == null) {
+            newNode = new Node<>(null, t, null);
+            first = newNode;
+        } else {
+            newNode = new Node<>(last, t, null);
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
+    }
+
+    private void addInTheMiddle(T t, int index) {
+        Node<T> newNode;
+        Node<T> currentNode = getNodeByIndex(index);
+        newNode = new Node<>(currentNode.prev, t, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode = first;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void checkIndex(int index, boolean isStrict) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("IndexOutOfBoundsException " + index);
+        }
+        if (!isStrict && index == size) {
+            throw new IndexOutOfBoundsException("IndexOutOfBoundsException " + index);
+        }
```

Comment:
I think we can combine these two 'if' blocks 
```suggestion
        if (index < 0 || index > size || !isStrict && index == size) {
            throw new IndexOutOfBoundsException("IndexOutOfBoundsException " + index);
        }
```